feat(components) adding bulleted list variables -I230

jerrybuks opened the pull request:
Signed-off-by: jerrybuks <profkiti@gmail.com>

# Issue #230 
Implemented the bulleted list variable as described in issue #230

### Changes
- modified ClauseComponent.js
- Added Clause variable Lists component

### Flags
- A user can now create  bulleted list variables

### Related Issues
N/A

dselman commented on src/components/ClauseComponent.js:
```diff
@@ -81,8 +86,35 @@ function ClauseComponent(props) {
 
+    let listData = {};
+    const findListVariables = (node) =>{
+      if(node.type === 'ul_list'){
```
What about 'ol_list' ?

jerrybuks replied:
ooh, I was thinking this feature was meant for only for "ul" since the issue was called "Bulleted List Variables", should I extend it to take care are of "ol" too?

mttrbrts commented on src/components/ClauseComponent.js:
```diff
@@ -20,6 +20,10 @@ import {
+import ClauseVariableList from './ClauseVariableList';
+
+/*slate-command */
+import regenerateKey from '../SlateCommands/RegenrateAllNodeKeys'
```
There is a typo in the file name, presumably this should be `RegenerateAllNodeKeys`

jerrybuks replied:
ooh, yeah, I didn't notice that, will fix it

irmerk commented on src/components/ClauseComponent.js:
```diff
@@ -81,8 +86,35 @@ function ClauseComponent(props) {
+            listData[listNode.key] = { 
+                currentText: listNode.text,
+                head: index === 0 ? true : false, 
+                tail: index === node.nodes.size - 1 ? true : false,
```
I see `head` and `tail` are used to determine if this is a node which can be removed or an empty node with an add icon. Can you explain though the logic here on line 96 and 97, I'm not sure I understand it fully.

jerrybuks replied:
Assuming there are several groups of variable lists in a clause, for each of these groups of lists, I needed to keep track of the first and last list item for each group, so while iterating through each list group, I keep track of the head by identifying the list item(a node) is at index 0, and keep track of the tail by identifying which list item is at the last node in each List group, by checking where index===" node.nodes.size - 1 " (i.e similar to saying array.length - 1). Therefore, in each list group, the head property of the first list item, will evaluate to true and the tail property of the last list item will evaluate to true. I now use these proeprties to ensure that the first list(head) can't be removed and the add icon, comes after the last node.

irmerk replied:
AH okay yes I get it now. Small code style difference but I think this could be:

```js
head: index === 0,
tail: index === (node.nodes.size - 1),
```

The ternary is redundant.

jerrybuks replied:
Gosh, my Bad, yeah!

irmerk commented on src/components/ClauseComponent.js:
```diff
@@ -111,11 +143,36 @@ function ClauseComponent(props) {
+    
+    let newNode = parentNode.nodes.asMutable();
+    
+    newNode =newNode.push(newBlockSlate);
```
Might not be possible, but could this be one line? 

```js
const newNode = parentNode.nodes.asMutable().push(newBlockSlate);
```

jerrybuks replied:
Yeah, it could, will change it

irmerk commented:
This looks SO good so far!

dselman commented:
Here are two templates: one uses an ol list, the other an ul list:

<img width="478" alt="Screenshot 2020-03-12 at 14 42 02" src="https://user-images.githubusercontent.com/623108/76533328-c9528f80-646f-11ea-9ad0-b5bbbfeea494.png">

jerrybuks commented:
> Here are two templates: one uses an ol list, the other an ul list:
> 
> <img alt="Screenshot 2020-03-12 at 14 42 02" width="478" src="https://user-images.githubusercontent.com/623108/76533328-c9528f80-646f-11ea-9ad0-b5bbbfeea494.png">

Yeah, I've added support for ordered lists too and added a sample to the demo

jeromesimeon commented:
> This is just awesome. Just a couple of things:
> 
> * It would be great to test on one of the volume discount + list templates Those would be either:
>   https://templates.accordproject.org/volumediscountolist@0.2.0.html or https://templates.accordproject.org/volumediscountulist@0.2.1.html and make sure the Clause text parses properly during edits (I think it gets printed on the console?)
> * I would really love to be able to add an item in the middle of the list (maybe using the "Return" key?) -- I don't think this should be a show stopper for this PR though.

This is a little bit clunky at the moment, but you can get sample text for those clauses to use in the demo (with wrapped variables) using:
`cicero normalize --wrapVariables` in those templates:

````
bash-3.2$ cd ~/git/cicero-template-library/src/volumediscountulist
bash-3.2$ cicero normalize --wrapVariables --output sampleWrapped.md
1:00:57 PM - info: Using current directory as template folder
1:00:57 PM - info: Loading a default text/sample.md file.
1:01:00 PM - info: Creating file: sampleWrapped.md
1:01:00 PM - info: Creating file: sampleWrapped.md
1:01:00 PM - info: Volume-Based Card Acceptance Agreement [Abbreviated]
----

This Agreement is by and between Card, Inc., a New York corporation, and you, the Merchant. By accepting the Card, you agree to be bound by the Agreement.
Discount means an amount that we charge you for accepting the Card, which amount is:
1. a percentage (Discount Rate) of the face amount of the Charge that you submit, or a flat per-
   Transaction fee, or a combination of both; and/or
1. a Monthly Flat Fee (if you meet our requirements).

Transaction Processing and Payments. Our Card acceptance, processing, and payment requirements are set forth in the Merchant Regulations. Some requirements are summarized here for ease of reference, but do not supersede the provisions in the Merchant Regulations.
Payment for Charges. We will pay you, through our agent, according to your payment plan in US dollars for the face amount of Charges submitted from your Establishments less all applicable deductions, rejections, and withholdings, which include:
1. the Discount,
1. any amounts you owe us or our Affiliates,
1. any amounts for which we have Chargebacks and
1. any Credits you submit.

Your initial Discount is indicated in the Agreement or otherwise provided to you in writing by us. In addition to your Discount we may charge you additional fees and assessments, as listed in the Merchant Regulations or as otherwise provided to you in writing by us. We may adjust any of these amounts and may change any other amount we charge you for accepting the Card.

### SETTLEMENT

#### Settlement Amount.

Our agent will pay you according to your payment plan, as described below, in US dollars for the face amount of Charges submitted from your Establishments less all applicable deductions, rejections, and withholdings, which include:
1. the Discount,
1. any amounts you owe us or our Affiliates,
1. any amounts for which we have Chargebacks, and
1. any Credits you submit.

Our agent will subtract the full amount of all applicable deductions, rejections, and withholdings, from this payment to you (or debit your Bank Account), but if it cannot, then you must pay it promptly upon demand.

#### Discount.

The Discount is determined according to the following table:

``` <list/>
- <variable id="volumeAbove" value="0.0"/>$ million <= Volume < <variable id="volumeUpTo" value="1.0"/>$ million : <variable id="rate" value="3.1"/>%
- <variable id="volumeAbove" value="1.0"/>$ million <= Volume < <variable id="volumeUpTo" value="10.0"/>$ million : <variable id="rate" value="3.1"/>%
- <variable id="volumeAbove" value="10.0"/>$ million <= Volume < <variable id="volumeUpTo" value="50.0"/>$ million : <variable id="rate" value="2.9"/>%
- <variable id="volumeAbove" value="50.0"/>$ million <= Volume < <variable id="volumeUpTo" value="500.0"/>$ million : <variable id="rate" value="2.5"/>%
- <variable id="volumeAbove" value="500.0"/>$ million <= Volume < <variable id="volumeUpTo" value="1000.0"/>$ million : <variable id="rate" value="1.2"/>%
- <variable id="volumeAbove" value="1000.0"/>$ million <= Volume < <variable id="volumeUpTo" value="1000000.0"/>$ million : <variable id="rate" value="0.1"/>%
```
````

irmerk commented:
@jerrybuks I think revert the `demo/src/index.js` file to normal and then add one of these (what @jeromesimeon and @dselman added) so the demo file has `Late Delivery and Penalty`, `Acceptance of Delivery`, and `Volume Discount ulist`

jerrybuks commented:
@irmerk, I have done as requested, just pushed the changes now and the parseManager works fine!

jerrybuks commented:
@jeromesimeon, I have just added the volumeDiscountList demo and the parseManger works well, I'm not sure I really understand how you want the "adding to the middle of the list" to work. would you mind explaining further

Michael-Grover commented:
Looks great, here's some of my feedback:

 - Remove the animation where the plus and minus icons slide when deleting and adding list items

jeromesimeon commented:
@jerrybuks Just tried the new version with the volume discount and did some edits to check the parsing results. it seems to work as expected! (not a small feat)

Only one snapshot but I added/removed lists, changed values, etc

<img width="1374" alt="Screenshot 2020-03-12 at 5 24 51 PM" src="https://user-images.githubusercontent.com/670099/76568177-b26b6780-6486-11ea-8458-c360420c5234.png">

irmerk commented:
> I would really love to be able to add an item in the middle of the list (maybe using the "Return" key?) -- I don't think this should be a show stopper for this PR though.

A nice to have, but if it's too much trouble we'll make that a separate issue and get this PR in.

> Remove the animation where the plus and minus icons slide when deleting and adding list items

This shouldn't stop this PR, but should be a quick fix with altering the `transition` css value?

> Allow the user to remove the first list item when there is more than one list item (but only when there is more than one list item)

Also a nice to have, but if it's too much trouble we'll make that a separate issue and get this PR in.

jeromesimeon commented:
> @jeromesimeon, I have just added the volumeDiscountList demo and the parseManger works well, I'm not sure I really understand how you want the "adding to the middle of the list" to work. would you mind explaining further

What I meant was: what would you do right now if you wanted to add an entry in your list in the middle of the list?

<img width="918" alt="Screenshot 2020-03-12 at 5 30 27 PM" src="https://user-images.githubusercontent.com/670099/76568621-9ae0ae80-6487-11ea-8d6f-a82f872ab558.png">

I _think_ right now the only way would be to delete all the following items and add them back? or did I miss something?

Michael-Grover commented:
[list icons.zip](https://github.com/accordproject/cicero-ui/files/4326732/list.icons.zip)
Let's replace the green icons with these

jeromesimeon commented:
> @jeromesimeon, I have just added the volumeDiscountList demo and the parseManger works well, I'm not sure I really understand how you want the "adding to the middle of the list" to work. would you mind explaining further

As an example, in a regular list I can press `return` (maybe this isn't the right approach design-wise here but to illustrate how it could be done)

![MiddleList](https://user-images.githubusercontent.com/670099/76569047-94066b80-6488-11ea-8955-33a10db39029.gif)

jerrybuks commented:
@irmerk @Michael-Grover @jeromesimeon I will make everyone's suggested changes and update the PR

jeromesimeon commented:
@jerrybuks 
> @irmerk @Michael-Grover @jeromesimeon I will make everyone's suggested changes and update the PR

For the "insert in the middle of the list" feature I maybe would check with @irmerk first. She might prefer a separate PR for this? I personally do not know how to evaluate the complexity and if it needs some design input from @Michael-Grover as well.

jerrybuks commented:
> @jerrybuks
> 
> > @irmerk @Michael-Grover @jeromesimeon I will make everyone's suggested changes and update the PR
> 
> For the "insert in the middle of the list" feature I maybe would check with @irmerk first. She might prefer a separate PR for this? I personally do not know how to evaluate the complexity and if it needs some design input from @Michael-Grover as well.

I have made all other changes except this, the reason is that it will most likely require an additional plugin because currently all the other plugins behavior in the codebase prevents going to a new line when "enter" is pressed and editor is locked. So, I guess there will be a need for an additional plugin to process adding a new variable list, when the cursor is at the end of a list and "enter key" is pressed. I do think, handling this in a separate PR is a good option.

jerrybuks commented:
> @jerrybuks
> 
> > @jeromesimeon, I have just added the volumeDiscountList demo and the parseManger works well, I'm not sure I really understand how you want the "adding to the middle of the list" to work. would you mind explaining further
> 
> What I meant was: what would you do right now if you wanted to add an entry in your list in the middle of the list?
> 
> <img alt="Screenshot 2020-03-12 at 5 30 27 PM" width="918" src="https://user-images.githubusercontent.com/670099/76568621-9ae0ae80-6487-11ea-8d6f-a82f872ab558.png">
> 
> I _think_ right now the only way would be to delete all the subsequent items and add them back? or did I miss something?
> 
> But I agree with @irmerk it isn't a show stopper this is a great step forward already. This could be discussed here after merging the PR or in a new issue.

There is a "not-perfect" option to this though, which is simply to edit the variables of the list at the point where you want to add a new item and maybe all the way down, till you get what you want.

irmerk commented on src/components/styles.js:
```diff
@@ -75,6 +80,9 @@ export const ClauseBody = styled.div`
     color: ${props => props.computedcolor || '#f1baff'};
   }
+  li{
+    transition:none;
```
I think I was mistaken, it must be `transition-duration` because they still don't instantly change position.

jerrybuks replied:
@irmerk I just tried it locally and it doesn't still make any difference. The "transition" is a shorthand to setting  "<property> <duration> <timing-function> <delay>", so setting the transition to none takes care of all the transition effects. I think this works though because if you adding or just removing the last element, it changes position instantly. The transition effect seems to happen when an element from any other position is being removed, of which I think it makes sense for it to work that way, so as to smoothly cover up the position of the element. So it seems like a default behavior and I'm not sure what to do to override it.

jeromesimeon commented:
> There is a "not-perfect" option to this though, which is simply to edit the variables of the list at the point where you want to add a new item and maybe all the way down, till you get what you want.

Agreed. This is also possibly time consuming and error prone, but this PR allows full list variable support in the sense you give it.

irmerk commented:
This is a bit unusual, but we're implementing a `development` branch to keep `master` more stable and for actual releases. Could you open a new PR but open it to merge this branch to `development` instead of `master`? You'll be all set and we can merge this in right away.

I'll be posting more information about this in slack and will be updating the `CONTRIBUTING` and `DEVELOPERS` guides soon.

jerrybuks commented:
Opened new PR to development branch as requested by @irmerk
